Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,479 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(Linux)
+
+import CNIOLinux
+
+// we stuff the event type into the user data for the sqe together with
+// the fd and sequenceIdentifier to match the events without needing any
+// memory allocations or references.
+
+internal enum CqeEventType : Int {
+    case poll = 1, pollModify, pollDelete // start with 1 to not get zero bit patterns for stdin
+}
+
+internal enum UringError: Error {
+    case loadFailure
+    case uringSetupFailure
+    case uringWaitCqeFailure
+}
+
+internal extension TimeAmount {
+    func kernelTimespec() -> __kernel_timespec {
+        var ts = __kernel_timespec()
+        ts.tv_sec = self.nanoseconds / 1_000_000_000
+        ts.tv_nsec = self.nanoseconds % 1_000_000_000
+        return ts
+    }
+}
+
+internal struct UringEvent {
+    var fd : Int32
+    var pollMask : UInt32
+    var sequenceIdentifier : UInt32
+}
+
+struct fdEventKey: Hashable {
+    var fileDescriptor : Int32
+    var sequenceIdentifier : UInt32
+
+    init(_ f: Int32, _ s : UInt32) {
+        self.fileDescriptor = f
+        self.sequenceIdentifier = s
+    }
+}
+
+final internal class Uring {
+    internal static let POLLIN: CUnsignedInt = numericCast(CNIOLinux.POLLIN)
+    internal static let POLLOUT: CUnsignedInt = numericCast(CNIOLinux.POLLOUT)
+    internal static let POLLERR: CUnsignedInt = numericCast(CNIOLinux.POLLERR)
+    internal static let POLLRDHUP: CUnsignedInt = numericCast(CNIOLinux.EPOLLRDHUP.rawValue) // FIXME: - POLLRDHUP not in ubuntu headers?!
```

Comment:
this seems odd that we use `POLL(IN|OUT|ERR|...)` but `EPOLLRDHUP`?

Reply:
Yeah, I'll go back investigate - for some reason is was not pulled in through the header on my machine, I'll sort it out.